config/translate.go

```diff
@@ -165,25 +180,53 @@ func TranslateFromV1(old v1.Config) types.Config {
 	return config
 }
 
+// golang--
+func translateV1MkfsOptionsToV2_1OptionSlice(opts v1.MkfsOptions) []types.Option {
+	newOpts := make([]types.Option, len(opts))
+	for i, o := range opts {
+		newOpts[i] = types.Option(o)
+	}
+	return newOpts
+}
+
+// golang--
+func translateStringSliceToV2_1SSHAuthorizedKeySlice(keys []string) []types.SSHAuthorizedKey {
+	newKeys := make([]types.SSHAuthorizedKey, len(keys))
+	for i, k := range keys {
+		newKeys[i] = types.SSHAuthorizedKey(k)
+	}
+	return newKeys
+}
+
+// golang--
+func translateStringSliceToV2_1UsercreateGroupSlice(groups []string) []types.UsercreateGroup {
+	var newGroups []types.UsercreateGroup
+	for _, g := range groups {
+		newGroups = append(newGroups, types.UsercreateGroup(g))
+	}
+	return newGroups
+}
+
 func TranslateFromV2_0(old v2_0.Config) types.Config {
 	translateVerification := func(old v2_0.Verification) types.Verification {
 		var ver types.Verification
 		if old.Hash != nil {
-			h := types.Hash(*old.Hash)
+			// .String() here is a wrapper around MarshalJSON, which will put the hash in quotes
+			h := strings.Trim(old.Hash.String(), "\"")
```

Comment:
This is going from the golang `Hash` type to a string of the format `<function>-<sum>`, which is necessary because the generated structs have the hash field as simply a string, whereas the non-generated structs in the v2.0 config can have the nicer-to-work-with `Hash` type.

The sum and function are both used in the result here, which hopefully answers both of your questions. An equivalent line would be `h := fmt.Sprintf("%s-%s", old.Hash.Function, old.Hash.Sum)`